File: 프로그래머스/lv2/42587. 프로세스/프로세스.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(vector<int> priorities, int location) {
    int answer = 0;
    priority_queue<int> pq;
    for(int i = 0; i < priorities.size(); i++){
        pq.push(priorities[i]);
    }
    
    queue<int> q;
    for(int i = 0; i < priorities.size(); i++) q.push(i);
    
    int idx = 0;
    while(!q.empty()) {
        int cur_location = q.front(); q.pop();
        if(priorities[cur_location] < pq.top()) {
            q.push(cur_location);
        }
        else{
            idx++;
            pq.pop();
            if(cur_location == location){
                answer = idx;
                break;
            }
        }
    }
    
    return answer;
}

File: 프로그래머스/2/42587. 프로세스/프로세스.py
def solution(priorities, location):
    stack = [(p, idx) for (idx, p) in enumerate(priorities)]
    
    answer = 0
    while 1:
        (curP, curIdx) = stack[0]
        (maxP, maxIdx) = max(stack, key = lambda x: x[0])
        
        if maxP > curP:
            stack.pop(0)
            stack.append((curP, curIdx))
        elif curIdx == location:
            answer += 1
            break
        else:
            answer += 1
            stack.pop(0)
            
    
    return answer